scylla/src/transport/locator/mod.rs

```diff
@@ -0,0 +1,554 @@
+            ))
+        }
+    }
+}
+
+#[derive(Debug)]
+enum ReplicaSetInner<'a> {
+    Plain(ReplicasArray<'a>),
+
+    // Represents a set of SimpleStrategy replicas that is limited to a specified datacenter.
+    FilteredSimple {
+        replicas: ReplicasArray<'a>,
+        datacenter: &'a str,
+    },
+
+    // Represents a set of NetworkTopologyStrategy replicas that is not limited to any specific
+    // datacenter. The set is constructed lazily, by invoking
+    // `locator.get_network_strategy_replicas()`.
+    ChainedNTS {
+        datacenter_repfactors: &'a HashMap<String, usize>,
+        locator: &'a ReplicaLocator,
+        token: Token,
+    },
+}
+
+/// Represents a set of replicas for a given token and strategy;
+///
+/// This container can only be created by calling `ReplicaLocator::replicas_for_token`, and it
+/// can borrow precomputed replica lists living in the locator.
+#[derive(Debug)]
+pub struct ReplicaSet<'a> {
+    inner: ReplicaSetInner<'a>,
+}
+
+impl<'a> ReplicaSet<'a> {
+    /// Chooses a random replica that satisfies the given predicate.
+    pub fn choose_filtered<R>(
+        self,
+        rng: &mut R,
+        predicate: impl Fn(NodeRef<'a>) -> bool,
+    ) -> Option<NodeRef<'a>>
+    where
+        R: Rng + ?Sized,
+    {
+        let happy = self.choose(rng)?;
+        if predicate(happy) {
+            return Some(happy);
+        }
+
+        self.into_iter()
+            .filter(|node_ref| predicate(node_ref))
+            .choose(rng)
+    }
+
+    /// Gets the size of the set.
+    ///
+    /// If the set represents `SimpleStrategy` replicas that were filtered by datacenter, this
+    /// function will have O(R) complexity, where R is the replication factor of that strategy.
+    ///
+    /// If the set represents `NetworkTopologyStrategy` replicas that were not filtered by
+    /// datacenter, this function will have O(D) complexity where D is the number of known
+    /// datacenters.
+    ///
+    /// In all other cases, the complexity is O(1)
+    pub fn len(&self) -> usize {
+        match &self.inner {
+            ReplicaSetInner::Plain(replicas) => replicas.len(),
+            ReplicaSetInner::FilteredSimple {
+                replicas,
+                datacenter,
+            } => replicas
+                .iter()
+                .filter(|node| node.datacenter.as_deref() == Some(datacenter))
+                .count(),
+            ReplicaSetInner::ChainedNTS {
+                datacenter_repfactors,
+                locator,
+                token: _,
+            } => datacenter_repfactors
+                .iter()
+                .map(|(dc, rf)| {
+                    let unique_nodes_in_dc_count = locator
+                        .unique_nodes_in_datacenter_ring(dc)
+                        .map(|nodes| nodes.len())
+                        .unwrap_or(0);
+                    dbg!(dc);
+                    dbg!(rf);
+                    dbg!(unique_nodes_in_dc_count);
```

Comment:
Please remove the debug prints.
They spam the console.

Reply:
Ups, I thought that I got rid of them sometime. Will fix.